### ISSUE TITLE
Patch prometheus container args with custom args storage.tsdb.max-block-duration / storage.tsdb.mix-block-duration

### ISSUE THREAD

**Issue by everythings-gonna-be-alright** [prometheus-operator/prometheus-operator#4414]
**What is missing?**

I want to change `storage.max-block-duration` / `storage.mix-block-duration` for reduce prometheus memory usage.

```
spec:
  containers:
  - args:
     - --storage.tsdb.min-block-duration=30m
     - --storage.tsdb.max-block-duration=30m
    name: prometheus
```
Now it replace all args for prometheus container.

**Why do we need it?**

To use `storage.tsdb.min-block-duration` / `storage.tsdb.max-block-duration` in prometheus without rewrite all args.
Like this solution: https://github.com/prometheus-operator/prometheus-operator/issues/3848

**Comment by paulfantom** [comment]
What's the use-case for changing those values to 30m?

**Comment by everythings-gonna-be-alright** [comment]
https://github.com/prometheus-operator/prometheus-operator/issues/3989

**Comment by everythings-gonna-be-alright** [comment]
I think it still actual for big clusters, when you collect big amount of metrics.
Or like my case:
I have 3 small clusters and 5GB of memory not enough for all my metrics. But i donʼt want provide more memory. It's insane.
storage.tsdb.max-block-duration=20m solve this problem.

**Comment by DmitriyStoyanov** [comment]
This feature needed to possibly fix issue prometheus/prometheus#6934 during prometheus startup.
Currently prometheus operator do not have ability to configure such flags.

We use kube-prometheus-stack to deploy prometheus in cluster and have issues with it described in linked issue.

**Comment by Boeller666** [comment]
I have the same problem. I use prometheus remote-write to send metrics from different clusters to an observer cluster running thanos-receive. As receive gets problems with big block-durations when a WAL replay is needed eg. after pod restarts, I need to decrease block-durations.

**Comment by stafot** [comment]
IMHO is a reasonable request for implementations with `thanos` as centralized point of truth. As the memory needs and the durations parameters could be different for each user of `prometheus-operator` depending to its scale.

**Comment by mzupan** [comment]
Battled this for a long time and this setup helped make progress. 

The reason we all want to config these is those that autoscale on a massive level you can easily have pod memory of 30-40gigs of memory.

You have to disable compaction for this to work. If you don't then it just ignores the arguments from what I can tell and keeps the default 2h in place. 

```
prometheus:
  prometheusSpec:
    containers:
      - name: prometheus
        args:
        - "--web.console.templates=/etc/prometheus/consoles"
        - "--web.console.libraries=/etc/prometheus/console_libraries"
        - "--config.file=/etc/prometheus/config_out/prometheus.env.yaml"
        - "--storage.tsdb.path=/prometheus"
        - "--web.enable-lifecycle"
        - "--storage.tsdb.no-lockfile"
        - "--web.enable-admin-api"
        - "--web.external-url=https://prometheus.fop.bar.com/"
        - "--web.route-prefix=/"
        - "--storage.tsdb.min-block-duration=10m"
        - "--storage.tsdb.max-block-duration=15m"
        - "--storage.tsdb.retention.size=800MB"
        - "--storage.tsdb.retention.time=1h"

    walCompression: false
```

I'm keeping only 800meg which in my testing is around 20-25min of metrics in wal files before they get checkpointed and deleted. I'm also staying steady around 7gig of memory used. For the first time in a long time though prometheus is reclaiming memory.

**Comment by simonpasquier** [comment]
With the latest version of the operator (v0.59.0), it should be possible to pass additional arguments via `spec.additionalArgs`.

```
kind: Prometheus
apiVersion: monitoring.coreos.com/v1 
...
spec:
  ...
  additionalArgs:
  - name: storage.tsdb.min-block-duration
    value: 30m
  - name: storage.tsdb.max-block-duration
    value: 24h
```

**Comment by SuperQ** [comment]
Thanks @simonpasquier, I was thinking about adding a configuration option for the DisableCompaction feature to allow overriding the hard-coded `2h` value there.

I'm looking to change our min/max block duration to 1h to help combat large (20M+ head series, 100Gi+ memory) instances with high pod churn rates. By compacting the head more often, we should be able to keep these large instances from building up too much head series memory during more frequent deploys or HPA changes.

**Comment by simonpasquier** [comment]
:wave: @SuperQ have you tested with the new `spec.additionalArgs` field as I indicated above?

**Comment by mhindery** [comment]
@simonpasquier  I am trying to set `additionalArgs` on the Prometheus CRD. I'm getting this validation error though everytime I try to update it:

```shell
# prometheuses.monitoring.coreos.com "kube-prometheus-kube-prome-prometheus" was not valid:
# * <nil>: Invalid value: "The edited file failed validation": ValidationError(Prometheus.spec): unknown field "additionalArgs" in com.coreos.monitoring.v1.Prometheus.spec
```

This is an environment from kube-prometheus using the latest chart (v40.1.0) which uses the prometheus operator v0.59.1 (quay.io/prometheus-operator/prometheus-operator:v0.59.1)

**Comment by simonpasquier** [comment]
@mhindery I guess that you haven't the latest version of the CRDs installed.

**Comment by hadican** [comment]
@mhindery You should have installed CRDs with:
```
kubectl apply --server-side -f manifests/setup
```

If you don't put `--server-side`, you might get error for annotations like: `Too long: must have at most 262144 bytes`

**Comment by frittentheke** [comment]
> Thanks @simonpasquier, I was thinking about adding a configuration option for the DisableCompaction feature to allow overriding the hard-coded `2h` value there.
> 
> I'm looking to change our min/max block duration to 1h to help combat large (20M+ head series, 100Gi+ memory) instances with high pod churn rates. By compacting the head more often, we should be able to keep these large instances from building up too much head series memory during more frequent deploys or HPA changes.

@SuperQ may I bluntly point at my issue / suggestion to have the storage management of Prometheus take the space required for compaction into account: https://github.com/prometheus/prometheus/issues/11112

While this would not free the need for tuning the settings (see comments like https://github.com/prometheus-operator/prometheus-operator/issues/4414#issuecomment-1090110995) but it would remove the risk of running out of space due to space for compaction not being taking into account.

**Comment by SuperQ** [comment]
It's not a question about running out of space. It's a question of reducing the memory footprint of the Prometheus memory used by head series when you have frequently series churn due to a large number of deployments that rotate the `instance` of every target.

@simonpasquier I started digging into this again, and no, `additionalArgs` won't help here because when running Thanos sicecars, `pkg/prometheus/statefulset.go` forcibly sets `disableCompaction` to true, which forcibly sets the the min/max block duration flags to `2h`. So even if I pass `disableCompaction: false` to in PrometheusSpec, it'll get flipped back on by the [Thanos check](https://github.com/prometheus-operator/prometheus-operator/blob/93cbae5583b30ee46f3094125411ee90261b482a/pkg/prometheus/statefulset.go#L819-L821).

A few ideas:
* Fix this by moving [`p.Spec.DisableCompaction`](https://github.com/prometheus-operator/prometheus-operator/blob/93cbae5583b30ee46f3094125411ee90261b482a/pkg/prometheus/statefulset.go#L717) to after Thanos setup.
* Add a "disableCompactionDuration" to the spec.

Happy to post a PR implementing one of these.

**Comment by simonpasquier** [comment]
@SuperQ the Thanos sidecar doc says:

> The --storage.tsdb.min-block-duration and --storage.tsdb.max-block-duration must be set to equal values to disable local compaction in order to use Thanos sidecar upload, otherwise leave local compaction on if sidecar just exposes StoreAPI and your retention is normal. The default of 2h is recommended. Mentioned parameters set to equal values disable the internal Prometheus compaction, which is needed to avoid the corruption of uploaded data when Thanos compactor does its job, this is critical for data consistency and should not be ignored if you plan to use Thanos compactor. Even though you set mentioned parameters equal, you might observe Prometheus internal metric prometheus_tsdb_compactions_total being incremented, don’t be confused by that: Prometheus writes initial head block to filesystem via its internal compaction mechanism, but if you have followed recommendations - data won’t be modified by Prometheus before the sidecar uploads it. Thanos sidecar will also check sanity of the flags set to Prometheus on the startup and log errors or warning if they have been configured improperly (#838).

So while the operator should enforce that min/max block duration are equal when the Thanos sidecar is configured for uploads, it should be possible for the user to define what's the value. IMHO the operator needs to look at the `additionalArgs` and if `--storage.tsdb.min-block-duration`  or `--storage.tsdb.max-block-duration`  is set, it should leave them untouched (e.g. let's assume that you know what you're doing). Otherwise it should set the default 2h duration for both (as currently). WDYT?

**Comment by gavol** [comment]
Same problem here: big cluster (thousands of workers) with thousands of PODs and not even 128 GB of RAM are enough.

I agree that is fine to enforce that min/max block duration should be equal when the Thanos sidecar is used; at the same time, the actual value should be configurable.

**Comment by SuperQ** [comment]
I think arg inspection is a bit unnecessary complication. I think a simple additional option to change the duration would be fine. Of course with some very strongly worded warnings that it could be a bad idea.

I can work on this.

**Comment by simonpasquier** [comment]
@SuperQ thanks for volunteering. I'm in as long as there's a clear "use at your own risks" statement.

**Comment by hakuno** [comment]
> With the latest version of the operator (v0.59.0), it should be possible to pass additional arguments via `spec.additionalArgs`.
> 
> ```
> kind: Prometheus
> apiVersion: monitoring.coreos.com/v1 
> ...
> spec:
>   ...
>   additionalArgs:
>   - name: storage.tsdb.min-block-duration
>     value: 30m
>   - name: storage.tsdb.max-block-duration
>     value: 24h
> ```

Additional args doesn't work.

```
level=error ts=2023-06-20T13:41:01.927083657Z caller=klog.go:116 component=k8s_client_runtime func=ErrorDepth msg="sync \"prometheus/prometheus\" failed: making statefulset failed: make StatefulSet spec: can't set arguments which are already managed by the operator: storage.tsdb.min-block-duration,storage.tsdb.max-block-duration"
```

**Comment by SuperQ** [comment]
@hakuno That is a separate issue, please create a new issue.

**Comment by pvlltvk** [comment]
@hakuno I'm not sure if it's still relevant for you, but you need to set  `disableCompaction` to `false` in order to use min-block-duration and max-block-duration via spec.additionalArgs

**Comment by hakuno** [comment]
@SuperQ @pvlltvk  Thank you both for commenting. In past, when I wrote that, those arguments were hardcoded settings. I don't know if something changed since then.

Anyway, I don't need it anymore. Thanks!

**Comment by guiguan** [comment]
you can set the following params to control min/max block duration:
- `disableCompaction: true`
- `thanos.blockSize: 1h`

then both `storage.tsdb.min-block-duration` and `storage.tsdb.max-block-duration` will be set to `1h` by the operator

**Comment by abhi4u1947** [comment]
1. This doesn't work because the operator code overrides it with a 2h value, even with disableCompaction: true.

```yaml
spec:
  containers:
    name: prometheus  
  - args:
     - --storage.tsdb.min-block-duration=30m
     - --storage.tsdb.max-block-duration=30m
```

2. This also doesn't work because the operator accepts the CR but gives an error in the operator logs:



```yaml
kind: Prometheus
apiVersion: monitoring.coreos.com/v1 
...
spec:
  ...
  disableCompaction: true
  additionalArgs:
  - name: storage.tsdb.min-block-duration
    value: 30m
  - name: storage.tsdb.max-block-duration
    value: 30m
```
Error:
```
msg="sync \"prometheus/prometheus\" failed: making statefulset failed: make StatefulSet spec: can't set arguments which are already managed by the operator: storage.tsdb.min-block-duration,storage.tsdb.max-block-duration"
```

3. Finally, this also can't be done, as we don't need Thanos Sidecar and with below minimum values it goes into crashloop, instead we prefer the remote write to Thanos Receiver endpoints.

```yaml
kind: Prometheus
apiVersion: monitoring.coreos.com/v1 
...
spec:
  ...
  disableCompaction: true
  thanos:
    blockSize: 30m
    objectStorageConfig:
      name: thanos-sidecar-objstore-config
      key: thanos.yaml
    listenLocal: true
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: thanos-sidecar-objstore-config
  namespace: monitoring
stringData:
  thanos.yaml: |
    {}
```

Basically there is no way to set the

**Comment by simonpasquier** [comment]
> Finally, this also can't be done, as we don't need Thanos Sidecar and with below minimum values it goes into crash loop, instead we prefer the remote write to Thanos Receiver endpoints.

have you tried the thanos sidecar without any object storage config?

```yaml
kind: Prometheus
apiVersion: monitoring.coreos.com/v1 
...
spec:
  ...
  disableCompaction: true
  thanos:
    blockSize: 30m
````

**Comment by Jakob3xD** [comment]
@simonpasquier It is possible to do this, but it starts the Pod with a Thanos sidecar, which is completely unneeded.

IMO This is still a valid use case when using remote write. We have a Prometheus just doing Blackbox Exporter things pushing the metrics to a receiver and I don't want the retention as it allocations memory I don't need.
Moreover, I can't scale on Memory usage even with sharding, as the Ram usage only decrease once the block is completed.
